src/util/CollectionUtils.ts

```diff
@@ -0,0 +1,18 @@
+/**
+ * Provides common functions for managing collections
+ */
+export class CollectionUtils {
+    /**
+     * Returns a dictionary of array values, indexed by property defind by `key`.
+     */
+    static indexBy<T>(list: T[], key: keyof T): {[key: string]: T} {
+        return list.reduce((obj, item) => {
+            const id = item[key] as any;
+            if (obj.hasOwnProperty(id)) {
+                throw new Error(`Key "${id}" is not unique!`);
+            }
+            obj[id] = item;
+            return obj;
+        }, {} as {[key: string]: T});
```

Comment:
I'm not a typescript expert, but a few points/questions:

1) isn't `[key: string]: T` a `Record<string, T>`? 
2) shouldn't `const id = item[key] as any` be number or string instead of `any`?

Reply:
I fixed 1). For 2) I'd rather have TypeScript check at compile time if the value of a specified key is a string or number and fail if it is not, but I'm not sure what's the best way to specify this. Do you have a suggestion?